rpc/client.go

```diff
@@ -91,219 +95,230 @@ type Client struct {
 }
 
-// IsHealthy returns whether the client is healthy.
-func (c *Client) IsHealthy() bool {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	return c.healthy
-}
-
-// Addr returns remote address of the client.
-func (c *Client) Addr() net.Addr {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	return c.addr
-}
+func (c *Client) setHealthy() {
+	if c.isHealthy {
+		return
+	}
 
-// LocalAddr returns the local address of the client.
-func (c *Client) LocalAddr() net.Addr {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	return c.lAddr
+	c.isHealthy = true
+	close(c.healthy.Load().(chan struct{}))
 }
 
-// RemoteOffset returns the most recently measured offset of the client clock
-// from the remote server clock.
-func (c *Client) RemoteOffset() proto.RemoteOffset {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	return c.offset
+func (c *Client) setUnhealthy() {
+	if c.isHealthy {
+		c.isHealthy = false
+		c.healthy.Store(make(chan struct{}))
+	}
 }
 
-// Close removes the client from the clients map and closes
-// the Closed channel.
+// Close closes the Closed channel, which triggers the end of the run loop and
+// removal from the clients map.
 func (c *Client) Close() {
 	clientMu.Lock()
-	if !c.closed {
-		if c.cached {
-			delete(clients, c.Addr().String())
-		}
-		c.mu.Lock()
-		c.healthy = false
-		c.closed = true
-		if c.Client != nil {
-			c.Client.Close()
-		}
-		c.mu.Unlock()
-		close(c.Closed)
+	defer clientMu.Unlock()
+
+	if c.isClosed {
+		return
 	}
-	clientMu.Unlock()
+
+	c.isClosed = true
+	delete(clients, c.key)
+	close(c.Closed)
 }
 
 // runHeartbeat sends periodic heartbeats to client. Closes the
 // connection on error. Heartbeats are sent in an infinite loop until
 // an error is encountered.
-func (c *Client) runHeartbeat(stopper *stop.Stopper) {
-	if log.V(2) {
-		log.Infof("client %s starting heartbeat", c.Addr())
-	}
-
+func (c *Client) runHeartbeat(retryOpts retry.Options) {
+	var err error
 	for {
+		for r := retry.Start(retryOpts); r.Next(); {
+			// reconnect unless the error is nil or definitely temporary
+			if netErr, ok := err.(net.Error); err != nil && !(ok && netErr.Temporary()) {
+				_ = c.connect()
+			}
+
+			err = c.heartbeat()
+
+			if err == nil {
+				c.setHealthy()
+				break
+			} else {
+				log.Warning(err)
+				c.setUnhealthy()
+			}
+		}
+		// Wait after the heartbeat so that the first iteration gets a wait-free
+		// heartbeat attempt.
 		select {
-		case <-stopper.ShouldStop():
+		case <-c.Closed:
 			return
 		case <-time.After(heartbeatInterval):
-			if err := c.heartbeat(); err != nil {
-				log.Infof("client %s heartbeat failed: %v; recycling...", c.Addr(), err)
-				return
-			}
 		}
 	}
 }
 
+// LocalAddr returns the local address of the client.
+func (c *Client) LocalAddr() net.Addr {
+	return util.MakeUnresolvedAddr(c.network, c.localAddress)
+}
+
+// RemoteAddr returns remote address of the client.
+func (c *Client) RemoteAddr() net.Addr {
+	return util.MakeUnresolvedAddr(c.network, c.remoteAddress)
```

Comment:
Our inputs for both the local and remote addresses are net.Addr objects; it seems a little silly to split those up into network/address pairs and then reconstruct them into "unresolved" addresses. `TCPAddr` has more than just two fields (for ipv6) so converting to/from an unresolved address involves constructing a string and parsing it.